core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java

```diff
@@ -387,12 +389,86 @@ public final HealthCheckedEndpointGroup build() {
                 LongMath.saturatedAdd(this.initialSelectionTimeoutMillis, delegate.selectionTimeoutMillis());
         final long selectionTimeoutMillis =
                 LongMath.saturatedAdd(this.selectionTimeoutMillis, delegate.selectionTimeoutMillis());
+        final SessionProtocol protocol = this.protocol;
+        final int port = this.port;
+        final Function<Endpoint, HealthCheckerParams> paramsFactory = endpoint -> {
+            final HealthCheckerParams delegate = paramsFactory().apply(endpoint);
+            return new AbstractHealthCheckerParams(delegate, protocol, port);
+        };
 
         return new HealthCheckedEndpointGroup(delegate, shouldAllowEmptyEndpoints(),
                                               initialSelectionTimeoutMillis, selectionTimeoutMillis,
-                                              protocol, port, retryBackoff,
-                                              clientOptionsBuilder.build(),
-                                              newCheckerFactory(), healthCheckStrategy);
+                                              retryBackoff, clientOptionsBuilder.build(),
+                                              newCheckerFactory(), healthCheckStrategy, paramsFactory);
+    }
+
+    static class AbstractHealthCheckerParams implements HealthCheckerParams {
+
+        private final HealthCheckerParams delegate;
+        private final SessionProtocol protocol;
+        private final int port;
+
+        AbstractHealthCheckerParams(HealthCheckerParams delegate, SessionProtocol protocol,
+                                    int port) {
+            this.delegate = delegate;
+            this.protocol = protocol;
+            this.port = port;
+        }
+
+        @Override
+        public String path() {
+            return delegate.path();
+        }
+
+        @Override
+        public HttpMethod httpMethod() {
+            return delegate.httpMethod();
+        }
+
+        @Override
+        @Nullable
+        public String host() {
+            return null;
+        }
+
+        @Override
+        public SessionProtocol protocol() {
+            return protocol;
+        }
+
+        @Override
+        public Endpoint endpoint() {
+            final Endpoint endpoint = delegate.endpoint();
+            if (port == 0) {
+                return endpoint.withoutDefaultPort(protocol);
+            } else if (port == protocol.defaultPort()) {
+                return endpoint.withoutPort();
+            } else {
+                return endpoint.withPort(port);
+            }
+        }
+    }
+
+    /**
+     * A dummy implementation to avoid breaking changes in {@link AbstractHealthCheckedEndpointGroupBuilder}.
+     */
+    Function<Endpoint, HealthCheckerParams> paramsFactory() {
+        return endpoint -> new HealthCheckerParamsAdapter() {
```

Comment:
What do you think about just always using the same `HealthCheckerParams` implementation class (or factory method) so we don't generate multiple classes? It looks like a simple value class.

Reply:
Sure, did a cast instead of the previous method chaining.